### ISSUE TITLE
Jzmo

### ISSUE THREAD

**Issue by Jzmo** [Jzmo/daily_report#1]

**Comment by Jzmo** [comment]
Date. Sep. 1st
### Progress
- Tasks
  - ASR on elder speech
- Problems:
  - Data scarcity: 
    - Deep learning approach usually requiring sufficient in-domain training data - 
  - Domain mismatch:
    - Most public speech training set are healthy speech.
  - Speaker mismatch:
    - Speech between elder speakers may have large differences due to health condition.
- Topics:
  - Data augmentation in data preparation stage – sp, specaug(in different level), GAN based method(good for domain adaption) 
  - Frontend: Speech enhancement for elder speech
  - In-domain Pretrain-finetune training:
    - **Iterative pseudo label training**
    - **Speaker adaption**
  - Out-of-domain Pretrain-finetune strategy
    - Commonly used Unsupervised/semi-supervised for pretrain stage
    - **Speaker adaption + domain adaption**
    - Domain mismatch estimation
      - t-SNE; Gaussian mean
    - **Model size and training time**
- Dataset:
  - LibriSpeech(pretrain), SwishBoard(10h subset for ft), Tedium(10h subset for ft), DementiaBank Pitt
- Baseline:
  - Supervised: hybrid DNN-HMM 
  - Semi/Unsupervised: wav2vec2.0, (finetune with in-domain data, e.g., DementiaBank Pitt train set)
- Uncertainty:
  - 
- Possible action items:
  - [ ] Data preprocessing, sp, remove long/short audio, vad based segment
  - [ ] Build and run baselines with 10h subset of noisy speech dataset and DementiaBank Pitt(supervised/finetune).
  - [ ] Uncertainty detection with embedding extracted from pretrained model, localize the error within utterances.
    - Could be done with small models.
Ongoing works:
1. IS2021 poster, tomorrow 10-12 pm, non-autoregressive special session
2. Reimplement Hubert[] in espnet, finished
3. Some optimization attempt for Hubert(discussion stage)

**Comment by Jzmo** [comment]
Date. Sep. 07th

### Progress
* Found some places to be fixed on Hubert-in-Espnet
   * unused lm related parametser
   * wrong name of train data folder(data/org/train -> data/train)
   * miss symbolic of LS data_prep.sh
   * reset GroupNorm bias initialization
* PR of Streaming MaskCTC
   * Made PR to yuya's repo. Thanks for your effort of reviewing it @yuya!
* Iteration1 Hubert is now being training, needs 4 days to finish
* Research ideas about hubert
   * Optimize training speed
       * larger frame shift of MFCC
       * One more subsample layer before Hubert
       * Running epochs - MFCC(10ms to 40 ms), Hubert(subsample rate from 320 to 640)
       * Both MFCC and Hubert label rate would be 25Hz(before was 100Hz and 50 Hz respectively)
   * Semi-supervised learning Hubert
       * Check the generalization ability of Hubert. 
       * Could evaluate pretraied model on nosiy data.
       * e.g. Pretrain on LS960 and finetune with 10hr SB subset.
       * (baseline)LS960 KM + LS960 Hubert + SB10hr finetune
       * SB10hr KM + LS960 Hubert + SB10hr finetune
       * SB10hr small supervised model(HMM-based or small model) + LS960 Hubert + SB10hr finetune
 ### Action items
- [ ] A new PR inclusing above items and README
- [ ] Continue run the lower label rate exp

**Comment by Jzmo** [comment]
Date. Sep. 08th

### Progress
- Conformer basedline for  Dementibank corpus:
  - Two augmentation compenations:
    - [Aug1]yezi's augmentation in (Development of the Cuhk Elderly Spee...)[https://ieeexplore.ieee.org/document/9413634],  {0.9, 1.0, 1.1}  +  {0.84, 0.95, 1.0, 1.08, 1.27}
    - [Aug2]espnet default {0.9, 1.0, 1.1} sp + specaug
  - Input feature: fbank
  - Model archtectore: elayer 8, dlayer 4, ahead 4, #units: 1024, conv-kernel: 7
  - MTL: 0.3ctc_loss+0.7att_loss
  - Result table(something wrong in the decoding):
  - https://docs.google.com/spreadsheets/d/1vF88epy11U0iWzABQ4iAOQ6fb-PZvqMKjt1p-A6Zjlw/edit?usp=sharing
- Collaboartion with Shinji:
  - Topic is optimize the training speed speech of Hubert model. 
  - The original version requires 32 gpus for 4days or 4 gpus for weeks without AMP
  - The plan is to make the model feasible for smaller number of gpus/shorter training time(ideally 1/10 training time), but keep the performance similar. 
  - Basic ideas is to dropout techniques(layerdrop, attention head drop...), and subsample layers. Still investigateing related methods.
  - In the start stage, frankly not have any results to report...

**Comment by Jzmo** [comment]
Date. Sep. 18th

### Progress
- Conformer basedline for  Dementibank corpus:
  - Word level output token(dict size = 3491) conformer
  - Smaller model: 6 encoder + 3 decoder
  - Add Specaug to yezi's ICCASP21 augmentation 
  - [Result table](https://docs.google.com/spreadsheets/d/1vF88epy11U0iWzABQ4iAOQ6fb-PZvqMKjt1p-A6Zjlw/edit?usp=sharing)
- Q:
  - TDNN train/validation wer gap? Conformer model CER (7/30) on (train/dev)
  - Any wrong transcription in dev/eval?
  - Toolkits about word-LM with char-ASR? Found flashlight(https://github.com/flashlight/flashlight), but failed in installation.
  - Can we use char-RNN-LM?

**Comment by Jzmo** [comment]
Date. e.g. Sep. 17
### Progress
* Optimizaion for hubert training speed: 
  * Subsample experiments:
    * x2, x4, x10
    * 
|          |                                       |            |                                       |   |
|----------|---------------------------------------|------------|---------------------------------------|---|
|          | Subsample rate of Hubert conv encoder | label rate | training time for 1 epoch with 4 gpus | valid acc mask on 20eps |
| Original | 320                                   | 50Hz       | 3.5hr                                 |  0.35 |
| x2       | 640                                   | 25Hz        | 2.5hr                                 |  0.29 |
| x4       | 1280                                  | 12.5Hz     | 2.0hr                                 |  0.109(stuck at this number) |
| x10      | 3200                                  | 5Hz        | 1.8hr                                 |  0.098(stuck at this number) |
  * Dropout test
    * Dropout heads - finish debuging, queue
    * Dropout layers - debugging stage
### Difficulties
* Almost impossible to run the 3 iterations for all experiments
* Hyperparameter need to be carelly tuned. 
  * For subsample setups - 
    * The label rates is lower than the original ones, which means model can see less information(noisy information) than the orginal ones. 
    * Mfcc pseudo label = 100Hz, and the label is ```[1,1,1,1,2,2,2,2,3,3,3,3]```
    * subsample rate = 320(Orignal Hubert setup), output label rate is 50 Hz, then the label in use is 
      * ```[1,**1**, 1,**1**,2,**2**,2,**2**,3,**3**,3,**3**]```
    * subsample rate = 640, output label rate is 25 Hz, then the label in use is 
      * ```[1,1,1,**1**,2,2,2,**2**,3,3,3,**3**]```
    * subsample rate = 1280, output label rate is 12.5 Hz, then the label in use is 
      * ```[1,1,1,1,2,2,2,**2**,3,3,3,3]```
    * Hyperparemter such as batchsize/accum-grad/lr may need to change in different setup to make it work.

**Comment by Jzmo** [comment]
Date. Sep. 28th

### Progress
- Conformer basedline for  Dementibank corpus:
  - char-LM weight weight 0.9 -> 0.7, test the effect
  - Fill the table with lexicon constrain/ word lm.
  - increase the beamsearch size for wordlm(maybe applying the word-level LM probability at each word end is too late
to recover better hypotheses)
  - [Result table](https://docs.google.com/spreadsheets/d/1vF88epy11U0iWzABQ4iAOQ6fb-PZvqMKjt1p-A6Zjlw/edit?usp=sharing)
  - Use lr=3 according to shujie's result.
- Q:
  - Multi-level LM? [https://www.merl.com/publications/docs/TR2017-181.pdf]
  - Apply current setup to shujie's torgo exp?
